design/kafka-upgrade/README.md

```diff
@@ -0,0 +1,175 @@
+
+## Changes to the `Kafka` CRD
+
+Add `Kafka.spec.kafka.version` to control which version of Kafka to use.
+This corresponds to using a different image for Kafka broker pods.
+The different image will embed a different Kafka distribution.
+
+The `Kafka.spec.kafka.version` will be a string matching the pattern
+`[0-9]+(\.[0-9]+)+`. It will _not_ be constrained by the CRD to the versions 
+supported by the operator since that would require coupling CRD versions to Kafka/Operator versions.
+
+From the user's point of view the upgrade procedure is:
+
+1. Edit the `Kafka` resource setting the `version` to the version being upgraded to
+   and setting `log.message.format.version` to the version being upgraded from.
+
+    ```
+    kind: Kafka
+    spec:
+      kafka:
+        version: 2.1.0
+        config:
+          log.message.format.version: 2.0.0
+          inter.broker.protocol.version: 2.0.0
+          # ...
+    ```
+
+    This will initiate a reconciliation which upgrades the pods but not the message format or interbroker version.
+
+2. Edit the `Kafka` resource upgrading or removing the `inter.broker.protocol.version`
+
+    ```
+    kind: Kafka
+    spec:
+      kafka:
+        version: 2.1.0
+        config:
+          log.message.format.version: 2.0.0
+          # ...
+    ```
+
+    This will initiate a rolling update which means the brokers are speaking the new protocol.
+    The message format for clients is still on the old version.
+
+3. The user now has to upgrade clients. This isn't something the CO can do.
+
+4. Once the user has upgraded the clients they can update (or remove) `log.message.format.version`:
+
+    ```
+    kind: Kafka
+    spec:
+      kafka:
+        version: 2.1.0
+        config:
+          # ...
+    ```
+
+    This will initiate a final reconciliation (rolling update). The upgrade is now complete.
+
+**Note:** It will be possible to revert the upgrade upto and including step 3.
+Because `log.message.format.version` functions as an assertion that all existing messages are on or below that message version, downgrading `log.message.format.version` is insufficient to guarantee that older clients don't see messages versions they don't understand.
+This is the main reason why the CO can't fully support version downgrade.
+
+## New images and changes to the build process
+
+Supporting multiple images requires changes to the build process.
+
+A different `kafka` image itself requires a different `kafka-base`.
+
+Consequently there would logically be different versions of the images for:
+
+* `kafka-connect`
+* `kafka-connect-s2i`
+* `zookeeper`
+* `kafka-mirror-maker`
+* `test-client`
+
+The way we identify images has to change too: We either use the tag to embed the Kafka version, or the name. 
+I.e. either
+
+  strimzi/kafka-base-2.0.0:latest
+  strimzi/kafka-base-2.0.0:0.9.0
+
+or 
+
+  strimzi/kafka-base:master-2.0.0
+  strimzi/kafka-base:0.9.0-2.0.0
```

Comment:
If the only thing that will be different between the images is the Kafka jar, my vote is for embedding the Kafka version in the tags, I think having all the versioning information in one place will make the images easier to manage. For one, the images with the same name will be grouped under the same images name in the Kubernetes and OpenShift registries, giving the appearance of less images. Also, i think it balances out a user's decision process, first choice is to just choose the image name, then worry about the versions later ( making it easier for users who just want the latest images and don't care about the versions). I think the people who are interested in the Kafka version will know the specific Kafka version they are looking for, so when they look another level deeper, from the image name to the tags, they will recognize the Kafka version in the tags next to the strimzi version.

Reply:
I am fine with either way, the main thing is having the information that other than the Kafka version is changed (2.0.0, 2.1.0 ...) so the JAR is changed, even something else could be changed in our `kafka-base` (i.e. scripts or commands to execute).